side_discontinuous option for Exodus output

roystgnr opened the pull request:
## Reason
MOOSE has had support for `SIDE_DISCONTINUOUS` variables for a while, but no way to visualize them easily.  This should fix that.

## Design
By setting side_discontinuous=true in Exodus output, disconnected fictitious side elements are added to that output, on which `SIDE_DISCONTINUOUS` values are plotted, with the discontinuities preserved.

A couple new DGKernel subclasses, HFEMTestJump and HFEMTrialJump, let us test this using a simple HFEM problem, solved with `SIDE_DISCONTINUOUS` lagrange multiplier rather than with the non-fictitious side element option previously available.

## Impact
Additional capabilities for Exodus.

DGKernels now "count" when our error checking tests to make sure that every variable in the system has an associated kernel.

Refs #20191.  It won't close that, though; we still have more 3D element types to add support for.


There's a bit of copy-and-paste code I still need to refactor into a new base class for SideObject+InternalSideObject, but I'm out of time for today and I'd like to find out whether CI agrees with my workstation that this is passing tests.

roystgnr committed:
Add HFEMTestJump/TrialJump DGKernels

This plus BCs should be all we need to do HFEM on an ordinary mesh
without a side elements block

roystgnr committed:
Side_discontinuous writing option for Exodus

Refs #20191

roystgnr committed:
Allow DGKernels to count as active for a var

With side-discontinuous variables they'll generally be the only kind of
kernel we use.

roystgnr committed:
Postprocessor to calculate L2 norm over elem sides

Previous HFEM tests were using the fact that side elements had been
constructed with their own block; with side-discontinuous elements those
don't exist and this is more useful.

roystgnr committed:
Test cases for Exodus SIDE_DISCONTINUOUS output

roystgnr committed:
Add documentation for new HFEM DGKernels

roystgnr committed:
Add doc for new postprocessor

roystgnr committed:
Add InternalSideUserObject::getFaceInfos()

This ought to be refactored, but my code is working now so I'm
committing it before I break it.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1195225/) on 0936a69 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22324/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22324/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 17f873bd9936bc936dab9e8d3356d693377b7c3e`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1195298/) on 36c2fd4 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22324/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1195331/) on 36c2fd4 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>17f873</th>
      <th colspan=4>#22324 36c2fd</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.29%</td>
      <td>84.28%</td>
      <td>-0.02%</td>
      <td>75.62%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>80109</td>
      <td>80230</td>
      <td>+121</td>
      <td>121</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14927</td>
      <td>14967</td>
      <td>+40</td>
      <td>39</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22324/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22324/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/richards)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22324/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 75.62% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad commented:
I think those nonsparse failures are likely real failures

roystgnr committed:
test for InternalSideIntegralVariablePostprocessor

I just intended this as an intermediate class, but it is usable by
itself too, and a test will also be good to point docs to.

roystgnr committed:
Document InternalSideIntegralVariablePostprocessor

roystgnr committed:
Fix typo

roystgnr commented:
> I think those nonsparse failures are likely real failures

After half a day of trying to figure this out: they're worse than real failures.  Real failures mean I wrote code that didn't do what I meant, and need to write code that does what I meant instead.  These failures are failures of *definition*, where I didn't actually have a straightforward idea of what I meant.

DGKernel calculations appear to use "which element has the higher id" to decide which side of an element counts as *the* element vs the mere *neighbor* ... which would mean that as soon as we add a few processors to a distributed run and DistributedMesh helpfully reshuffles ids to give us contiguous id blocks on each rank, the definition of "element" vs "neighbor" inverts, and the definition of a Lagrange multiplier variable constraining directional jumps from element to neighbor also inverts, and exodiff starts whining at me about having rightfully noticed that `-1/6` and `1/6` are two different numbers.

I guess the right thing to do here is something like a `vertex_average()`-sorting-dependent sign flip in the new HFEM kernels, to make the results depend on geometry without regard for numbering.

I bet the old HFEM kernels with the non-fictitious side elements have the same problem, but never trigger it because they don't run distributed...

roystgnr committed:
Make HFEM lambda side variable geometry-defined

This should fix inconsistencies we see when a numbering-defined variable
is compared from run to run with partitioning-dependent numberings.

lindsayad commented:
I would contend that difference only by sign in a Lagrange multiplier is not a true failure at all. What I would do is create a custom comparison that doesn't exodiff based on the LMs ... but then again I guess the whole point of this PR is the exodus output of these LMs :laughing: Would it be possible to pass the LM field through an aux kernel that takes the absolute value of the LM and then only write out that field?

roystgnr commented:
> I guess the whole point of this PR is the exodus output of these LMs

Admittedly I do feel like I've gone down the rabbit hole here.  I haven't bothered to push anything until it's all fixed, but as part of fixing those doc failures I've written a test case to add to the documentation of a base class I refactored to properly support the postprocessor I wrote for a *different* test case ...

But I do want this stuff to be useful too, and if I can get this working (my quick fix idea seems to have worked perfectly in 1D but not at all in 2D???) then making it easier for future kernels to be tested via exodiff will be a valuable thing to have too.

Plus I really want to understand what I'm still doing wrong.

roystgnr commented:
Okay, I get it now.  After making the orientation of lambda depend on geometry independent of numbering, I now get consistent lambda values on all side elements, and I was already getting consistent u values on all interiors ... but I'm also plotting u on those side elements.  And in this test u is discontinuous.  And I plot its value on a side exactly once, from whichever interior I happen to hit first when I'm plotting that side.  So if renumbering means I hit interiors in different orders, then I can plot the same internal side from different elements in different partitionings, and it can take on different values.

Technically u is discontinuous, IMHO it would be fair to just plot it as 0 or NaN on sides ... but users might find it more useful to plot it as the average of the values between the two interiors, and that's pretty much what we already do when plotting discontinuous fields at shared nodes in other contexts, too.

lindsayad commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -3340,6 +3340,7 @@ NonlinearSystemBase::checkKernelCoverage(const std::set<SubdomainID> & mesh_subd
   global_kernels_exist |= _nodal_kernels.hasActiveObjects();
 
   _kernels.subdomainsCovered(input_subdomains, kernel_variables);
+  _dg_kernels.subdomainsCovered(input_subdomains, kernel_variables);
```
nice

lindsayad commented on test/tests/outputs/exodus/exodus_side_discontinuous.i:
```diff
@@ -0,0 +1,88 @@
+[Executioner]
+  type = Steady
+  solve_type = 'NEWTON'
+  petsc_options_iname = '-pc_type -snes_linesearch_type -pc_factor_mat_solver_type'
+  petsc_options_value = 'lu       basic                 mumps'
```
Not saying you need to but would be cool to use a field split here

lindsayad reviewed (changes requested):
I'd like to see a `PetscJacobianTester` test to verify that the Jacobian for this higher-d-variable/side-variable coupling is accurate

roystgnr committed:
Add HFEM Jacobian test

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1236181/) on 1c15426 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22324/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22324/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format b4abe9f2057135c183fbc01a98ff639246b75c65`

roystgnr committed:
clang-format

lindsayad commented:
Is there a way to hit `InternalSideUserObject::getFaceInfos`?

lindsayad commented on framework/doc/content/source/postprocessors/ElementSidesL2Norm.md:
```diff
@@ -0,0 +1,35 @@
+!equation
+||u||_{L_2(\Gamma)} = \sqrt{\sum_{\text{mesh element sides}} \sum_{\text{quadrature point qp}} J[qp] W[qp] c[qp] u(qp)^2}
+
+These terms are initialized on eeach side, so the $J$ term accounts
```
```suggestion
These terms are initialized on each side, so the $J$ term accounts
```

roystgnr committed:
Update framework/doc/content/source/postprocessors/ElementSidesL2Norm.md

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

lindsayad commented:
Yea probably a new distinct test would be the easiest. Let me know if you'd like any help cooking one up

roystgnr committed:
Test to cover InternalSideUserObject::getFaceInfos

roystgnr commented:
And of course now I need an issue tag to add to the test that I added to cover the framework feature that I added to cover the postprocessor that I added to test the kernels that I added to test the output I added to visualize the library feature I added to swallow the cat to catch the bird to catch the spider to catch the fly...

lindsayad commented:
That was surprisingly straightforward to follow up to the cat :laughing: :grimacing: :cry:

lindsayad reviewed (approved):
Thanks for all the hard work on this
